File: ui/src/types/requestResult.ts
import { List } from "immutable";

export type RequestResult = {
    statusCode: StatusCode | null;
    validationFailures: List<ValidationFailure> | null;
    exception: Exception | null;
    permissionFailure: PermissionFailure | null;
};

export enum StatusCode {
    validationError = 442,
    PermissionFailure = 403,
    ok = 200,
    internalServerError = 500,
}

export type ValidationFailure = {
    propertyName: string;
    errorMessage: string;
    attemptedValue: any;
};

export type PermissionFailure = {
    failureMessage: string;
    underlyingRight: string;
};

export type Exception = {
    message: string;
    innerException: any;
};

File: API/BLL/Base/RequestResult.cs
using System;
using System.Collections.Generic;
using FluentValidation.Results;

namespace API.BLL.Base
{
    public class RequestResult
    {
        public IList<ValidationFailure> ValidationFailures { get; set; }
        public PermissionFailure PermissionFailure { get; set; }
        public StatusCode StatusCode { get; set; }
        public Exception Exception { get; set; }

        public RequestResult()
        {
        }

        public RequestResult(RequestResult existing)
        {
            ValidationFailures = existing.ValidationFailures;
            PermissionFailure = existing.PermissionFailure;
            StatusCode = existing.StatusCode;
            Exception = existing.Exception;
        }
    }

    public enum StatusCode
    {
        Unauthorized = 401,
        PermissionFailure = 403,
        ValidationError = 442,
        Ok = 200,
        InternalServerError = 500
    }
}